Lib/test/_test_multiprocessing.py

```diff
@@ -5420,6 +5421,219 @@ def test__all__(self):
+        with self.connection.Listener() as l:
+            with self.connection.Client(l.address) as c:
+                with l.accept() as d:
+                    c.send(1729)
+                    self.assertEqual(d.recv(), 1729)
+
+        if self.TYPE == 'processes':
+            self.assertRaises(OSError, l.accept)
+
+        self.assertEqual(self.spy['loads'], 1)
+        self.assertEqual(self.spy['dump'], 1)
+
+    @classmethod
+    def _put_and_get_in_queue(cls, queue, parent_can_continue):
+        parent_can_continue.set()
+        queue.put("Something")
+        queue.get(timeout=TIMEOUT2)
+        close_queue(queue)
+        parent_can_continue.set()
+
+    @unittest.skipUnless(HAS_REDUCTION, "test needs multiprocessing.reduction")
+    def test_queue_custom_reducer(self):
+        queue = self.Queue()
+        parent_can_continue = self.Event()
+        p = self.Process(target=self._put_and_get_in_queue,
+                         args=(queue, parent_can_continue))
+        p.start()
+        parent_can_continue.wait()
+        element = queue.get(timeout=TIMEOUT3)
+        self.assertEqual(element, "Something")
+        queue.put("Other_Something")
+        parent_can_continue.wait()
+        close_queue(queue)
+        p.join(timeout=TIMEOUT3)
+        p.terminate()
+        self.assertEqual(element, "Something")
+        self.assertEqual(self.spy['loads'], 1)
+        self.assertEqual(self.spy['dump'], 3)
+        self.assertEqual(p.exitcode, 0)
+
+    @unittest.skipUnless(HAS_REDUCTION, "test needs multiprocessing.reduction")
+    def test_listener_custom_reduction(self):
+        with self.connection.Listener() as l:
+            with self.connection.Client(l.address) as c:
+                with l.accept() as d:
+                    c.send(1729)
+                    self.assertEqual(d.recv(), 1729)
+
+        if self.TYPE == 'processes':
+            self.assertRaises(OSError, l.accept)
+
+        self.assertEqual(self.spy['loads'], 1)
+        self.assertEqual(self.spy['dump'], 1)
+
+class CustomContext(multiprocessing.context.BaseContext):
+    _name = "custom"
+    Process = multiprocessing.get_context().Process
+
+
+class _TestCustomReducerWithContext(BaseTestCase):
+    """Test case for per-context reducers"""
+
+    ALLOWED_TYPES = ('processes',)
+
+    def setUp(self):
+        self.custom_ctx = CustomContext()
+        self.default_ctx = multiprocessing.get_context()
+        self.original_custom_reducer = self.custom_ctx.get_reducer()
+        self.original_default_reducer = self.default_ctx.get_reducer()
+        self.default_spy = {'loads': 0, 'dump': 0}
+        self.custom_spy = {'loads': 0, 'dump': 0}
+
+    def tearDown(self):
+        self.custom_ctx.set_reducer(self.original_custom_reducer)
+        self.default_ctx.set_reducer(self.original_default_reducer)
+
+    @classmethod
+    def _put_and_get_in_queue(cls, queue, parent_can_continue):
+        parent_can_continue.set()
```

Comment:
If you want to re-use the event afterwards, you have to clear it at some point. But I'm not sure all this synchronization is necessary (the queue already synchronizes for you).